docs/vfio-user.rst

```diff
@@ -1131,6 +1144,168 @@ client must write data on the same order and transction size as read.
+round-trip through the client.
+
+The server returns an array describing sub-regions of the given region along
+with the server specifies a set of sub-regions and the requested file descriptor
+notification mechanism to use for that sub-region.  Each sub-region in the
+response message may choose to use a different method, as defined below.  The
+two mechanisms supported in this specification are ioeventfds and ioregionfds.
+
+A client should then hook up the returned file descriptors with the notification
+method requested.
+
+Region IO FD info format
+^^^^^^^^^^^^^^^^^^^^^^^^
+
++------------+--------+------+
+| Name       | Offset | Size |
++============+========+======+
+| argsz      | 16     | 4    |
++------------+--------+------+
+| flags      | 20     | 4    |
++------------+--------+------+
+| index      | 24     | 4    |
++------------+--------+------+
+| count      | 28     | 4    |
++------------+--------+------+
+| subregions | 32     | ...  |
++------------+--------+------+
+
+* *argsz* is the size of the region IO FD info structure plus the
+  total size of the subregion array. Thus, each array entry "i" is at offset
+  i * ((argsz - 32) / count). Note that currently this is 40 bytes for both IO
+  FD types, but this is not to be relied on.
+* *flags* must be zero
+* *index* is the index of memory region being queried
+* *count* is the number of sub-regions in the array
+* *subregions* is the array of Sub-Region IO FD info structures
+
+The client must set ``flags`` to zero and specify the region being queried in
+the ``index``.
+
+The client sets the ``argsz`` field to indicate the maximum size of the response
+that the server can send, which must be at least the size of the response header
+plus space for the subregion array. If the full response size exceeds ``argsz``,
+then the server must respond only with the response header and the Region IO FD
+info structure, setting in ``argsz`` the buffer size required to store the full
+response. In this case, no file descriptors are passed back.  The client then
+retries the operation with a larger receive buffer.
+
+The reply message will additionally include at least one file descriptor in the
+ancillary data. Note that more than one subregion may share the same file
+descriptor.
+
+Each sub-region given in the response has one of two possible structures,
+depending whether *type* is `VFIO_USER_IO_FD_TYPE_IOEVENTFD` or
+`VFIO_USER_IO_FD_TYPE_IOREGIONFD`:
+
+Sub-Region IO FD info format (ioeventfd)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
++-----------+--------+------+
+| Name      | Offset | Size |
++===========+========+======+
+| offset    | 0      | 8    |
++-----------+--------+------+
+| size      | 8      | 8    |
++-----------+--------+------+
+| fd_index  | 16     | 4    |
++-----------+--------+------+
+| type      | 20     | 4    |
++-----------+--------+------+
+| flags     | 24     | 4    |
++-----------+--------+------+
+| padding   | 28     | 4    |
```

Comment:
Why do we need padding? I think we can have a short explanation in the form of a `.. Note::`.

Reply:
It's not unusual to pad data to natural alignment, doesn't seem worth commenting on...